### DISCUSSION TITLE
How to get hydrogen indices in substruct matches

### DISCUSSION THREAD

**Discussion by cinematikk** [rdkit/rdkit#5289] [Q&A] [answered]
I am searching for functional groups in RDKit molecules and want to get the corresponding atom indices. Using `GetSubstructMatches()` however does not return some (implicit?) hydrogen indices. My code is

```
from rdkit import Chem

patt = Chem.MolFromSmarts("[CH3;X4]")
patt = Chem.AddHs(patt, explicitOnly=True)

mol = Chem.MolFromSmiles("CC(=O)Br")
mol = Chem.AddHs(mol)

match_indices = mol.GetSubstructMatches(patt)
```
This only returns the index of the carbon atom. How could I get the hydrogen indices as well?

**Comment by bp-kelley** [comment, marked as answer]
The query you are looking for is a Carbon that has three hydrogens and four connections, you aren't explicitly looking for the actual hydrogens in the match graph.  To do this, they need to be proper atoms in the query.  Unfortunately, AddHs doesn't work on the smarts patterns since valence rules on smarts don't make a lot of sense and the RDKit certainly doesn't understand the query "three hydrogens with four connections" in its valence rules.  

To properly do this, you need to add explicit  [H] query atoms to the smarts pattern as follows:

```
patt = Chem.MolFromSmarts("[CH3;X4]([H])([H])[H]")
mol = Chem.MolFromSmiles("CC(=O)Br")
molh = Chem.AddHs(mol)

print(len(molh.GetSubstructMatch(patt))) # this is the expected 4
```

**Reply by cinematikk** [reply]
Thanks a lot!
